packages/SwingSet/src/kernel/virtualObjectManager.js

```diff
@@ -0,0 +1,342 @@
+import { assert, details, q } from '@agoric/assert';
+import { parseVatSlot } from '../parseVatSlots';
+
+export function makeVirtualObjectManager(
+  syscall,
+  allocateExportID,
+  valToSlotTable,
+  m,
+  cacheSize,
+) {
+  /**
+   * Make a simple LRU cache of virtual object inner selves.
+   *
+   * @param {number} size  Maximum number of entries to keep in the cache before
+   *    starting to throw them away.
+   */
+  function makeCache(size) {
+    let lruHead;
+    let lruTail;
+    let count = 0;
+    const liveTable = new Map();
+
+    const cache = {
+      makeRoom() {
+        while (count > size && lruTail) {
+          liveTable.delete(lruTail.instanceKey);
+          lruTail.dropData();
+          if (lruTail.prev) {
+            lruTail.prev.next = undefined;
+          } else {
+            lruHead = undefined;
+          }
+          lruTail = lruTail.prev;
+          count -= 1;
+        }
+      },
+      flush() {
+        const saveSize = size;
+        size = 0;
+        cache.makeRoom();
+        size = saveSize;
+      },
+      remember(innerObj) {
+        if (liveTable.has(innerObj.instanceKey)) {
+          return;
+        }
+        liveTable.set(innerObj.instanceKey, innerObj);
+        innerObj.prev = undefined;
+        cache.makeRoom();
+        innerObj.next = lruHead;
+        if (lruHead) {
+          lruHead.prev = innerObj;
+        }
+        lruHead = innerObj;
+        if (!lruTail) {
+          lruTail = innerObj;
+        }
+        count += 1;
+      },
+      refresh(innerObj) {
+        if (innerObj !== lruHead) {
+          const oldPrev = innerObj.prev;
+          const oldNext = innerObj.next;
+          if (oldPrev) {
+            oldPrev.next = oldNext;
+          } else {
+            lruHead = oldNext;
+          }
+          if (oldNext) {
+            oldNext.prev = oldPrev;
+          } else {
+            lruTail = oldPrev;
+          }
+          innerObj.prev = undefined;
+          innerObj.next = lruHead;
+          lruHead.prev = innerObj;
+          lruHead = innerObj;
+        }
+      },
+      lookup(instanceKey) {
+        return liveTable.get(instanceKey);
+      },
+    };
+    return cache;
+  }
+
+  const cache = makeCache(cacheSize);
+
+  /**
+   * Map from virtual object kind IDs to reanimator functions for the
+   * corresponding kinds of virtual objects.
+   */
+  const kindTable = new Map();
+
+  /**
+   * Produce a representative given an instance key.  Used for deserializing.
+   *
+   * @param {string} vref  The instanceID of the object being dereferenced
+   *
+   * @returns {*}  A representative of the object identified by `vref`
```

Comment:
I think we can be more precise (though untested):

```suggestion
   * @returns {Object}  A representative of the object identified by `vref`
```

Reply:
I don't know if this is right, but the type checker seems to think it's ok.